File: c++/sort/SortColors.cpp
//
//  SortColors.cpp
//  c++
//
//  Created by sure on 1/20/15.
//  Copyright (c) 2015 Shuo Li. All rights reserved.
//
// Mistake was made: when i == blue + 1, then left of the array are all 2s,
// No need to continue any more
// Know sth new: std::swap
// Thanks for the hint from: http://fisherlei.blogspot.com/2013/01/leetcode-sort-colors.html

#include <stdio.h>
#include <iostream>

class Solution {
public:
    void sortColors(int A[], int n) {
        int red = 0;
        int blue = n - 1;
        for (int i = 0; i < n; i++) {
            if (i == blue + 1) {
                break;
            }
            if (A[i] == 0) {
                std::swap(A[red], A[i]);
                red ++;
                continue;
            }
            if (A[i] == 2) {
                std::swap(A[blue], A[i]);
                blue --;
                i --;
            }
        }
    }
};

int main() {
    Solution *s = new Solution();
    int A[1] = {2};
    s->sortColors(A, 1);
    for (int i = 0; i < sizeof(A)/sizeof(*A); i++) {
        std::cout<<A[i]<<std::endl;
    }
    
    delete s;

    return 0;
}

File: java/SortColors.java
public class TwoPassSolution {
    public void sortColors(int[] A) {
        int zero = this.moveZeroToLeft(A);
        this.moveTwoToRight(A, zero);
    }
    
    public void swap(int[] A, int slow, int fast) {
        int temp = A[slow];
        A[slow] = A[fast];
        A[fast] = temp;
    }
    
    public int moveZeroToLeft(int[] A) {
        int slow = 0;
        int fast = 0;
        while (fast < A.length) {
            if (A[fast] == 0) {
                this.swap(A, slow, fast);
                slow += 1;
            }
            fast += 1;
        }
        return slow;
    }
    
    public void moveTwoToRight(int[] A, int zero) {
        int slow = A.length - 1;
        int fast = A.length - 1;
        while (fast >= zero) {
            if (A[fast] == 2) {
                this.swap(A, slow, fast);
                slow -= 1;
            }
            fast -= 1;
        }
    }
}

// My previous solution is actually two passes.
// Here's one pass solution from:
// http://fisherlei.blogspot.com/2013/01/leetcode-sort-colors.html
public class OnePassSolution {
    public void sortColors(int[] nums) {
        int red = 0;
        int blue = nums.length - 1;
        int i = 0;
        while (i < blue + 1) {
            if (nums[i] == 1) {
                i ++;
            } else if (nums[i] == 2) {
                swap(nums, i, blue);
                blue --;
            } else {
                swap(nums, i, red);
                red ++;
                i ++;
            }
        }
    }
    
    public void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}
